### PR TITLE
Adding -Ofast for sample build.

### PR DESCRIPTION
**Pull request by ggfan**
re-open this one as I deleted the branch  for https://github.com/google/oboe/pull/164 accidently

### CODE REVIEW

**Comment by ggfan** [comment]
added flag OBOE_ENABLE_LOGGING for https://github.com/google/oboe/issues/175: @dturner  please really review it.  Thanks!

File: samples/debug-utils/logging_macros.h [hunk 1]
```diff
@@ -18,7 +18,7 @@
 #define __SAMPLE_ANDROID_DEBUG_H__
 #include <android/log.h>
 
-#if 1
+#if OBOE_ENABLE_LOGGING
```
Comment (philburk) [inline samples/debug-utils/logging_macros.h]: Why is this change here. It conflicts with an older fix.

File: samples/MegaDrone/CMakeLists.txt [hunk 2]
```diff
@@ -27,7 +27,8 @@ target_link_libraries( megadrone log oboe )
...
 
-# Append compiler optimization flags to both the debug and release build variants
-set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -Ofast")
-set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -Ofast")
-
+# Enable optimization for release build; it could be enabled for debug build
+# if performance is of interest.
+target_compile_options(megadrone
+    PRIVATE
+        -std=c++14 -Wall $<$<CONFIG:RELEASE>:-Ofast>)
```
Comment (dturner) [inline samples/MegaDrone/CMakeLists.txt]: Why add `-std=c++14`? I don't think the app needs any C++14 features in which case it makes sense to leave this out.

File: CMakeLists.txt [hunk 3]
```diff
@@ -27,8 +27,12 @@ target_include_directories(oboe
...
 
-target_compile_options(oboe PRIVATE -std=c++11
+ # Compile Flags:
+ #     Enable -Werror when building debug config
+ #     Enable -Ofast for release config (could be enabled for debug too)
+ target_compile_options(oboe PRIVATE -std=c++11
                             PRIVATE -Wall
-                            PRIVATE "$<$<CONFIG:DEBUG>:-Werror>") # Only include -Werror when building debug config
+                            PRIVATE "$<$<CONFIG:DEBUG>:-Werror>"
+                            PRIVATE "$<$<CONFIG:RELEASE>:-Ofast>")
```
Comment (dturner) [inline CMakeLists.txt]: Oboe is currently built from source when a dependent project (e.g. LiveEffect) is built. Is the build configuration (release or debug) used when building Oboe?

File: samples/LiveEffect/src/main/cpp/CMakeLists.txt [hunk 4]
```diff
@@ -39,6 +39,8 @@ target_link_libraries(liveEffect
         atomic
         log)
 
+# Enable optimization for release build; it could be enabled for debug build
+# if performance is of interest.
 target_compile_options(liveEffect
     PRIVATE
-        -Wall -Werror)
+        -std=c++14 -Wall -Werror $<$<CONFIG:RELEASE>:-Ofast>)
```
Comment (dturner) [inline samples/LiveEffect/src/main/cpp/CMakeLists.txt]: Why add -std=c++14? I don't think the app needs any C++14 features in which case it makes sense to leave this out.

File: samples/MegaDrone/CMakeLists.txt (hunk 2 above)
Comment (dturner) [reply samples/MegaDrone/CMakeLists.txt]: Also, for this particular example `-Ofast` should be enabled for both debug and release because it's a performance related sample and we don't want first-time developers thinking there's a bug in the code because of audio glitches in the default (debug) configuration.

Comment (ggfan) [reply samples/MegaDrone/CMakeLists.txt]: I thought we have email said that the standard is to use C++14 as a standard configuration, and enable -Ofast for release.

Comment (ggfan) [reply samples/MegaDrone/CMakeLists.txt]: [clang -g doc](https://clang.llvm.org/docs/CommandGuide/clang.html#code-generation-options) points out that -g  works best with -o0.  Implying -ofast enabled, source level debugging may have trouble.  I think to enable it for release; for debugging version, leave it out and leave users to add it.  thanks

File: samples/debug-utils/logging_macros.h (hunk 1 above)
Comment (ggfan) [reply samples/debug-utils/logging_macros.h]: this is for this: https://github.com/google/oboe/issues/175. 

add a flag to guard it, then could use compile flag to control it is on or off

File: samples/MegaDrone/CMakeLists.txt (hunk 2 above)
Comment (dturner) [reply samples/MegaDrone/CMakeLists.txt]: I agree that `-O0` should be used in normal circumstances but for the MegaDrone sample it will almost certainly glitch and hence should be left as `-Ofast` for all variants.

File: CMakeLists.txt [hunk 5]
```diff
@@ -27,7 +27,15 @@ target_include_directories(oboe
...
                            PUBLIC include)
 
+ # Compile Flags:
+ #     Enable -Werror when building debug config
+ #     Enable -Ofast
 target_compile_options(oboe PRIVATE -std=c++11
                             PRIVATE -Wall
-                            PRIVATE "$<$<CONFIG:DEBUG>:-Werror>") # Only include -Werror when building debug config
+                            PRIVATE "$<$<CONFIG:DEBUG>:-Werror>"
+                            PRIVATE -Ofast)
```
Comment (dturner) [inline CMakeLists.txt]: Shouldn't this be release=`-Ofast`, debug=`-O0`? 
Also why the multiple `PRIVATE` keywords? Couldn't these all go on one line?

File: CMakeLists.txt [hunk 6]
```diff
@@ -27,7 +27,15 @@ target_include_directories(oboe
...
+
+# Enable logging when debugging
+# target_compile_options(oboe PUBLIC -DOBOE_ENABLE_LOGGING=1)
```
Comment (dturner) [inline CMakeLists.txt]: Set `OBOE_ENABLE_LOGGING=1` for debug configuration and not for release

File: samples/LiveEffect/src/main/cpp/CMakeLists.txt [hunk 7]
```diff
@@ -39,6 +39,7 @@ target_link_libraries(liveEffect
         atomic
         log)
 
-target_compile_options(liveEffect
-    PRIVATE
-        -Wall -Werror)
+# Enable optimization flags: if having problems with source level debugging,
+# disable -Ofast ( and debug ), re-enable it after done debugging.
+target_compile_options(liveEffect PRIVATE -Wall -Werror -Ofast)
```
Comment (dturner) [inline samples/LiveEffect/src/main/cpp/CMakeLists.txt]: Again only -Ofast for release config

File: samples/MegaDrone/CMakeLists.txt [hunk 8]
```diff
@@ -27,7 +27,6 @@ target_link_libraries( megadrone log oboe )
 # Make the Oboe public headers available to our app
 include_directories (${OBOE_DIR}/include)
 
-# Append compiler optimization flags to both the debug and release build variants
-set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -Ofast")
-set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -Ofast")
-
+# Enable optimization flags: if having problems with source level debugging,
+# disable -Ofast ( and debug ), re-enable it after done debugging.
+target_compile_options(megadrone PRIVATE -Wall -Werror -Ofast)
```
Comment (dturner) [inline samples/MegaDrone/CMakeLists.txt]: -Ofast for release only

File: samples/RhythmGame/CMakeLists.txt [hunk 9]
```diff
@@ -42,4 +42,11 @@ set (OBOE_DIR ../..)
...
 
 # Specify the path to the Oboe header files.
-include_directories (${OBOE_DIR}/include)
\ No newline at end of file
+include_directories (${OBOE_DIR}/include)
+
+# Enable optimization flags: if having problems with source level debugging,
+# disable -Ofast ( and debug ), re-enable after done debugging.
+target_compile_options(native-lib
+    PRIVATE -std=c++14
```
Comment (dturner) [inline samples/RhythmGame/CMakeLists.txt]: Multiple PRIVATE statements

File: samples/hello-oboe/CMakeLists.txt [hunk 10]
```diff
@@ -53,3 +53,7 @@ add_library(hello-oboe SHARED
 
 # Specify the libraries needed for hello-oboe
 target_link_libraries(hello-oboe android log oboe)
+
+# Enable optimization flags: if having problems with source level debugging,
+# disable -Ofast ( and debug ), re-enable after done debugging.
+target_compile_options(hello-oboe PRIVATE -std=c++14 -Wall -Werror -Ofast)
```
Comment (dturner) [inline samples/hello-oboe/CMakeLists.txt]: I think the only reason for using C++14 is that the sample is using `std::make_unique` which I believe it isn't.

File: samples/MegaDrone/CMakeLists.txt (hunk 2 above)
Comment (ggfan) [reply samples/MegaDrone/CMakeLists.txt]: done with next checkin, thx

File: samples/LiveEffect/src/main/cpp/CMakeLists.txt (hunk 4 above)
Comment (ggfan) [reply samples/LiveEffect/src/main/cpp/CMakeLists.txt]: done

File: CMakeLists.txt (hunk 5 above)
Comment (ggfan) [reply CMakeLists.txt]: add -Ofast for debug, leave default for debug build. Done

File: CMakeLists.txt (hunk 6 above)
Comment (ggfan) [reply CMakeLists.txt]: restore its original for this one.

File: samples/LiveEffect/src/main/cpp/CMakeLists.txt (hunk 7 above)
Comment (ggfan) [reply samples/LiveEffect/src/main/cpp/CMakeLists.txt]: done

File: samples/MegaDrone/CMakeLists.txt (hunk 8 above)
Comment (ggfan) [reply samples/MegaDrone/CMakeLists.txt]: first review, you said:
"I agree that -O0 should be used in normal circumstances but for the MegaDrone sample it will almost certainly glitch and hence should be left as -Ofast for all variants."
the same file but conflicting comment from you, which one you are going?

File: samples/RhythmGame/CMakeLists.txt (hunk 9 above)
Comment (ggfan) [reply samples/RhythmGame/CMakeLists.txt]: done

File: samples/hello-oboe/CMakeLists.txt (hunk 10 above)
Comment (ggfan) [reply samples/hello-oboe/CMakeLists.txt]: done

**Comment by ggfan** [comment]
I think the code review requests have been addressed with this PR, please review again.

left -Ofast for release and debug builds for Oboe library as performance has higher priority than source code level debugging, assuming most users would trace to app code, not oboe lib source code. Please double check, it is all just assumption.

Thanks!

File: samples/MegaDrone/CMakeLists.txt (hunk 8 above)
Comment (dturner) [reply samples/MegaDrone/CMakeLists.txt]: My apologies Gerry, this comment was meant for the other samples. Here's what the optimisation flag should be for each sample: 

hello-oboe: debug -O0, release -Ofast
LiveEffect: debug -O0, release -Ofast
RhythmGame: debug -O0, release -Ofast
MegaDrone: all -Ofast
